### ISSUE TITLE
Spurious warnings from deprecation macro in RooFit

### ISSUE THREAD

**Issue by andresailer** [root-project/root#15150]
### Check duplicate issues.

- [X] Checked for duplicates

### Description

Just including RooDataSet leads to spurious warnings
```
/cvmfs/sft-nightlies.cern.ch/lcg/nightlies/dev3/Fri/ROOT/HEAD/x86_64-el9-gcc12-opt/include/ROOT/RConfig.hxx:500:73: warning: 'REMOVE_THIS_NOW' attribute directive ignored [-Wattributes]
  500 | #define _R_DEPRECATED_REMOVE_NOW(REASON) __attribute__((REMOVE_THIS_NOW))
      |                                                                         ^
/cvmfs/sft-nightlies.cern.ch/lcg/nightlies/dev3/Fri/ROOT/HEAD/x86_64-el9-gcc12-opt/include/ROOT/RConfig.hxx:508:37: note: in expansion of macro '_R_DEPRECATED_REMOVE_NOW'
  508 | # define _R__DEPRECATED_634(REASON) _R_DEPRECATED_REMOVE_NOW(REASON)
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~
/cvmfs/sft-nightlies.cern.ch/lcg/nightlies/dev3/Fri/ROOT/HEAD/x86_64-el9-gcc12-opt/include/ROOT/RConfig.hxx:460:39: note: in expansion of macro '_R__DEPRECATED_634'
  460 | #   define _NAME3_(name1,name2,name3) name1##name2##name3
      |                                       ^~~~~
/cvmfs/sft-nightlies.cern.ch/lcg/nightlies/dev3/Fri/ROOT/HEAD/x86_64-el9-gcc12-opt/include/ROOT/RConfig.hxx:477:30: note: in expansion of macro '_NAME3_'
  477 | #   define _R__JOIN3_(F,X,Y) _NAME3_(F,X,Y)
      |                              ^~~~~~~
/cvmfs/sft-nightlies.cern.ch/lcg/nightlies/dev3/Fri/ROOT/HEAD/x86_64-el9-gcc12-opt/include/ROOT/RConfig.hxx:522:3: note: in expansion of macro '_R__JOIN3_'
  522 |   _R__JOIN3_(_R__DEPRECATED_,MAJOR,MINOR)("will be removed in ROOT v" #MAJOR "." #MINOR ": " REASON)
      |   ^~~~~~~~~~
/cvmfs/sft-nightlies.cern.ch/lcg/nightlies/dev3/Fri/ROOT/HEAD/x86_64-el9-gcc12-opt/include/RooDataSet.h:93:6: note: in expansion of macro 'R__DEPRECATED'
   93 |      R__DEPRECATED(6,34, "Use RooDataSet(name, title, vars, Import(*tree), Cut(cutVar), WeightVar(wgtVarName)).");
      |      ^~~~~~~~~~~~~
```

### Reproducer

```term
cat > main.cpp << EOF
#include <RooDataSet.h>
int main() {
  return 0;
}
EOF
source /cvmfs/sft-nightlies.cern.ch/lcg/nightlies/dev3/Fri/ROOT/HEAD/x86_64-el9-gcc12-opt/ROOT-env.sh
g++ -I $ROOTSYS/include main.cpp  -o a.out
```

### ROOT version

```
$ root --version
ROOT Version: 6.33/01
Built for linuxx8664gcc on Apr 04 2024, 22:54:25
From heads/master@v6-31-01-1587-gaf0e1abffd
```

### Installation method

build from sources in lcgcmake, options https://gitlab.cern.ch/sft/lcgcmake/-/blob/master/projects/CMakeLists.txt#L95

### Operating system

alma9, alma8, ubuntu20, ubuntu22

### Additional context

This breaks packages that are not careful about `-isystem` and use `-Werror`, other are just flooded with useless warnings.

**Comment by dpiparo** [comment]
Hi @andresailer , thanks for reporting. It's not spurious but rather by design. We are converging towards a solution: see https://github.com/root-project/root/pull/15148 .

**Comment by andresailer** [comment]
I am sorry to say, but I think your design needs a revision.
```
cat > sane.hh << EOF
#pragma once

[[deprecated]]
int deprecatedFunction() {
  return 123;
}
int saneFunction() {
  return 321;
}
EOF
cat > sane.cpp << EOF
#include "sane.hh"
int main() {
  saneFunction();
  return 0;
}
EOF
cat > depr.cpp << EOF
#include "sane.hh"
int main() {
  deprecatedFunction();
  return 0;
}
EOF
g++ depr.cpp
g++ sane.cpp
```
One only gets a warning when deprecatedFunction is called.

**Comment by pcanal** [comment]
> One only gets a warning when deprecatedFunction is called.

Yes. The idea of the set of macros is (for a function that we no exist in v6.34.00 for example:

1. Warn in v6.32.xy that the function is deprecated only for code where the function is called.
2. Warn as soon as we start the development cycle of v6.34 that the function (used or not) that we said were going to be removed are removed.

2. is currently implemented by turning it in v6.33.01 (i.e. the first day of the v6.34 development cycles.  https://github.com/root-project/root/pull/15148  postpone this to the next development tag (6.33.02).

As you point out, this system however make it difficult to use the master (eg. with `-Werror`) until we do remove the functions.  (Side note, prior to this automatic system we routinely 'forgot' to remove deprecated function when intended)

**Comment by guitargeek** [comment]
Thanks for chiming in @pcanal!

Indeed, this is fixed for now by https://github.com/root-project/root/pull/15148.

I'll mark the fixed version as "not applicable", because this is not relevant for any release.
